--- openssh-6.6p1/openbsd-compat/fmt_scaled.c~	2014-05-13 17:23:35.999776940 +0300
+++ openssh-6.6p1/openbsd-compat/fmt_scaled.c	2014-05-13 17:23:40.989697403 +0300
@@ -49,6 +49,11 @@
 #include <ctype.h>
 #include <limits.h>
 
+#if !defined(LLONG_MAX)
+#   define LLONG_MAX	9223372036854775807LL
+#   define LLONG_MIN	(-LLONG_MAX - 1LL)
+#endif
+
 typedef enum {
 	NONE = 0, KILO = 1, MEGA = 2, GIGA = 3, TERA = 4, PETA = 5, EXA = 6
 } unit_type;
--- openssh-6.6p1/openbsd-compat/strtonum.c~	2006-08-05 09:27:20.000000000 +0300
+++ openssh-6.6p1/openbsd-compat/strtonum.c	2014-05-13 17:31:27.715592099 +0300
@@ -26,6 +26,11 @@
 #include <limits.h>
 #include <errno.h>
 
+#if !defined(LLONG_MAX)
+#   define LLONG_MAX   9223372036854775807LL
+#   define LLONG_MIN   (-LLONG_MAX - 1LL)
+#endif
+
 #define INVALID 	1
 #define TOOSMALL 	2
 #define TOOLARGE 	3
