--- openssh-1.2.3/Makefile.in.orig	Thu Apr  6 19:31:03 2000
+++ openssh-1.2.3/Makefile.in	Thu Apr  6 19:31:52 2000
@@ -23,6 +23,7 @@
 PATHS=-DETCDIR=\"$(sysconfdir)\" -DSSH_PROGRAM=\"$(SSH_PROGRAM)\" -DSSH_ASKPASS_DEFAULT=\"$(ASKPASS_PROGRAM)\"
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -59,7 +60,7 @@
 	$(LD) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh $(LIBS)
 
 sshd: libssh.a	$(SSHDOBJS)
-	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh $(LIBS)
+	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh $(LIBWRAP) $(LIBS) 
 
 scp: libssh.a scp.o
 	$(LD) -o $@ scp.o $(LDFLAGS) -lssh $(LIBS) 
--- openssh-1.2.3/configure.in.orig	Thu Apr  6 19:32:46 2000
+++ openssh-1.2.3/configure.in	Thu Apr  6 19:33:25 2000
@@ -735,6 +735,9 @@
 				[
 					AC_MSG_RESULT(yes)
 					AC_DEFINE(LIBWRAP)
+					LIBWRAP="-lwrap"
+					AC_SUBST(LIBWRAP)
+					LIBS="$saved_LIBS"
 				],
 				[
 					AC_MSG_RESULT(no)
