diff -urN openssh-3.0p1.orig/configure.ac openssh-3.0p1/configure.ac
--- openssh-3.0p1.orig/configure.ac	Sat Nov  3 20:09:33 2001
+++ openssh-3.0p1/configure.ac	Wed Nov  7 16:22:54 2001
@@ -123,6 +123,8 @@
 	AC_DEFINE(LOCKED_PASSWD_PREFIX, "!!")
 	AC_DEFINE(SPT_TYPE,SPT_REUSEARGV)
 	inet6_default_4in6=yes
+	no_libnsl=1
+	no_libsocket=1
 	case `uname -r` in
 	1.*|2.0.*)
 		AC_DEFINE(BROKEN_CMSG_TYPE)
--- openssh-3.2.3p1/configure.ac.orig	Sat May 25 13:02:18 2002
+++ openssh-3.2.3p1/configure.ac	Sat May 25 13:14:58 2002
@@ -360,7 +359,6 @@
 	util.h utime.h utmp.h utmpx.h)
 
 # Checks for libraries.
-AC_CHECK_FUNC(yp_match, , AC_CHECK_LIB(nsl, yp_match))
 AC_CHECK_FUNC(setsockopt, , AC_CHECK_LIB(socket, setsockopt))
 
 dnl SCO OS3 needs this for libwrap
