--- openssh-10.4p1.orig/servconf.c	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/servconf.c	2026-07-06 21:23:34.091821111 +0200
@@ -1185,6 +1185,11 @@ process_server_config_line_depth(ServerO
 			*charptr = xstrdup(arg);
 		break;
 #endif
+#ifdef CHROOT
+	case sUseChroot:
+		intptr = &options->use_chroot;
+		goto parse_flag;
+#endif
 
 	/* Standard Options */
 	case sBadOption:
--- openssh-10.4p1.orig/servconf.h	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/servconf.h	2026-07-06 21:23:28.301821115 +0200
@@ -268,6 +268,7 @@ SSHCONF_ALIAS(KeepAlive, TCPKeepAlive, S
 	SSHD_CONFIG_ENTRIES_LEGACY \
 	SSHD_CONFIG_ENTRIES_ALIASES \
 	SSHD_CONFIG_ENTRIES_PAM \
+	SSHD_CONFIG_ENTRIES_CHROOT \
 	SSHD_CONFIG_ENTRIES_LASTLOG
 
 #ifdef USE_PAM
@@ -280,6 +281,14 @@ SSHCONF_UNSUPPORTED_INT(use_pam, UsePAM,
 SSHCONF_UNSUPPORTED_STRING(pam_service_name, PAMServiceName, SSHCFG_GLOBAL)
 #endif
 
+#ifdef CHROOT
+#define SSHD_CONFIG_ENTRIES_CHROOT \
+SSHCONF_INTFLAG(use_chroot, UseChroot, SSHCFG_GLOBAL, 0, SSHCFG_COPY_NONE)
+#else
+#define SSHD_CONFIG_ENTRIES_CHROOT \
+SSHCONF_UNSUPPORTED_INT(use_chroot, UseChroot, SSHCFG_GLOBAL)
+#endif
+
 #ifdef DISABLE_LASTLOG
 #define SSHD_CONFIG_ENTRIES_LASTLOG \
 SSHCONF_UNSUPPORTED_INT(print_lastlog, PrintLastLog, SSHCFG_GLOBAL)
--- openssh-10.4p1.orig/session.c	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/session.c	2026-07-06 21:22:56.644136453 +0200
@@ -1316,6 +1316,10 @@ void
 do_setusercontext(struct passwd *pw)
 {
 	char uidstr[32], *chroot_path, *tmp;
+#ifdef CHROOT
+	char *user_dir;
+	char *new_root;
+#endif /* CHROOT */
 
 	platform_setusercontext(pw);
 
@@ -1358,6 +1362,29 @@ do_setusercontext(struct passwd *pw)
 			free(options.chroot_directory);
 			options.chroot_directory = NULL;
 			in_chroot = 1;
+#ifdef CHROOT
+		} else if (!in_chroot && options.use_chroot) {
+			user_dir = xstrdup(pw->pw_dir);
+			new_root = user_dir + 1;
+
+			while ((new_root = strchr(new_root, '.')) != NULL) {
+				new_root--;
+				if (strncmp(new_root, "/./", 3) == 0) {
+					*new_root = '\0';
+					new_root += 2;
+
+					if (chroot(user_dir) != 0)
+						fatal("Couldn't chroot to user directory %s", user_dir);
+					/* NOTE: session->pw comes from pwcopy(), so replace pw_dir this way (incompatible with plain getpwnam() or getpwnam_r()) */
+					free(pw->pw_dir);
+					pw->pw_dir = xstrdup(new_root);
+					in_chroot = 1;
+					break;
+				}
+				new_root += 2;
+			}
+			free(user_dir);
+#endif /* CHROOT */
 		}
 
 #ifdef HAVE_LOGIN_CAP
--- openssh-10.4p1.orig/sshd_config	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/sshd_config	2026-07-06 21:22:56.644685847 +0200
@@ -83,6 +83,10 @@ AuthorizedKeysFile	.ssh/authorized_keys
 # and KbdInteractiveAuthentication to 'no'.
 #UsePAM no
 
+# Set this to 'yes' to enable support for chrooted user environment.
+# You must create such environment before you can use this feature.
+#UseChroot yes
+
 #AllowAgentForwarding yes
 #AllowTcpForwarding yes
 #GatewayPorts no
--- openssh-10.4p1.orig/sshd_config.0	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/sshd_config.0	2026-07-06 21:22:56.645435161 +0200
@@ -1309,6 +1309,16 @@ DESCRIPTION
              open channels.  This option may be useful in conjunction with
              ChannelTimeout.
 
+     UseChroot
+             Specifies whether to use chroot-jail environment with ssh/sftp,
+             i.e. restrict users to a particular area in the filesystem. This
+             is done by setting user home directory to, for example,
+             /path/to/chroot/./home/username.  sshd looks for a '.' in the
+             users home directory, then calls chroot(2) to whatever directory
+             was before the . and continues with the normal ssh functionality.
+             For this to work properly you have to create special chroot-jail
+             environment in a /path/to/chroot directory.
+
      UseDNS  Specifies whether sshd(8) should look up the remote host name,
              and to check that the resolved host name for the remote IP
              address maps back to the very same IP address.
--- openssh-10.4p1.orig/sshd_config.5	2026-07-06 09:57:12.000000000 +0200
+++ openssh-10.4p1/sshd_config.5	2026-07-06 21:22:56.646363281 +0200
@@ -1982,6 +1982,16 @@ Gives the facility code that is used whe
 The possible values are: DAEMON, USER, AUTH, LOCAL0, LOCAL1, LOCAL2,
 LOCAL3, LOCAL4, LOCAL5, LOCAL6, LOCAL7.
 The default is AUTH.
+.It Cm UseChroot
+Specifies whether to use chroot-jail environment with ssh/sftp, i.e. restrict
+users to a particular area in the filesystem. This is done by setting user
+home directory to, for example, /path/to/chroot/./home/username.
+.Nm sshd
+looks for a '.' in the users home directory, then calls
+.Xr chroot 2
+to whatever directory was before the . and continues with the normal ssh
+functionality. For this to work properly you have to create special chroot-jail
+environment in a /path/to/chroot directory.
 .It Cm TCPKeepAlive
 Specifies whether the system should send TCP keepalive messages to the
 other side.
