diff -urNp -x '*.orig' openssh-8.4p1.org/sandbox-seccomp-filter.c openssh-8.4p1/sandbox-seccomp-filter.c
--- openssh-8.4p1.org/sandbox-seccomp-filter.c	2020-09-27 09:25:01.000000000 +0200
+++ openssh-8.4p1/sandbox-seccomp-filter.c	2021-03-01 11:30:52.381809172 +0100
@@ -204,6 +204,9 @@ static const struct sock_filter preauth_
 #ifdef __NR_futex
 	SC_ALLOW(__NR_futex),
 #endif
+#ifdef	__NR_futex
+	SC_ALLOW(__NR_futex),
+#endif
 #ifdef __NR_geteuid
 	SC_ALLOW(__NR_geteuid),
 #endif
