--- a/configure.ac~	2017-10-02 21:34:26.000000000 +0200
+++ b/configure.ac	2017-10-06 08:35:16.756080761 +0200
@@ -1487,7 +1487,7 @@ AC_ARG_WITH(ldns,
 	ldns=""
 	if test "x$withval" = "xyes" ; then
 		AC_PATH_TOOL([LDNSCONFIG], [ldns-config], [no])
-		if test "x$PKGCONFIG" = "xno"; then
+		if test "x$LDNSCONFIG" = "xno"; then
 			CPPFLAGS="$CPPFLAGS -I${withval}/include"
 			LDFLAGS="$LDFLAGS -L${withval}/lib"
 			LIBS="-lldns $LIBS"

