diff -urN openssh-3.9p1.org/configure.ac openssh-3.9p1/configure.ac
--- openssh-3.9p1.org/configure.ac	2004-08-17 19:03:29.098600648 +0200
+++ openssh-3.9p1/configure.ac	2004-08-17 19:06:43.675020536 +0200
@@ -2266,12 +2266,12 @@
 				       [ AC_MSG_RESULT(yes)
 					 AC_DEFINE(HEIMDAL)
 					 K5LIBS="-lkrb5 -ldes"
-					 K5LIBS="$K5LIBS -lcom_err -lasn1"
+					 K5LIBS="$K5LIBS -lasn1"
 					 AC_CHECK_LIB(roken, net_write, 
 					   [K5LIBS="$K5LIBS -lroken"])
 				       ],
 				       [ AC_MSG_RESULT(no)
-					 K5LIBS="-lkrb5 -lk5crypto -lcom_err"
+					 K5LIBS="-lkrb5 -lk5crypto"
 				       ]
 			)
 			AC_SEARCH_LIBS(dn_expand, resolv)

