diff -bruN openssh-2.1.0/Makefile.in openssh-2.1.0.new/Makefile.in
--- openssh-2.1.0/Makefile.in	Tue May  9 06:28:55 2000
+++ openssh-2.1.0.new/Makefile.in	Wed May 10 19:16:45 2000
@@ -23,6 +23,7 @@
 PATHS=-DETCDIR=\"$(sysconfdir)\" -DSSH_PROGRAM=\"$(SSH_PROGRAM)\" -DSSH_ASKPASS_DEFAULT=\"$(ASKPASS_PROGRAM)\"
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -71,7 +72,7 @@
 	$(LD) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
 
 sshd: libssh.a	libopenbsd-compat.a $(SSHDOBJS)
-	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
+	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) $(LIBWRAP)
 
 scp: libopenbsd-compat.a libssh.a scp.o
 	$(LD) -o $@ scp.o $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) 
diff -bruN openssh-2.1.0/configure.in openssh-2.1.0.new/configure.in
--- openssh-2.1.0/configure.in	Mon May  8 12:49:37 2000
+++ openssh-2.1.0.new/configure.in	Wed May 10 20:44:06 2000
@@ -756,8 +756,11 @@
 	[  --with-tcp-wrappers     Enable tcpwrappers support],
 	[
 		if test "x$withval" != "xno" ; then
+			# we need to add -lwrap to LIBS only for configure
+			# testing, and then remove
 			saved_LIBS="$LIBS"
-			LIBS="$LIBS -lwrap"
+			LIBWRAP="-lwrap"
+			LIBS="$LIBS $LIBWRAP"
 			AC_MSG_CHECKING(for libwrap)
 			AC_TRY_LINK(
 				[
@@ -768,6 +771,7 @@
 				[
 					AC_MSG_RESULT(yes)
 					AC_DEFINE(LIBWRAP)
+					AC_SUBST(LIBWRAP)
 				],
 				[
 					AC_MSG_RESULT(no)
@@ -775,6 +779,7 @@
 					LIBS="$saved_LIBS"
 				]
 			)
+			LIBS="$saved_LIBS"
 		fi
 	]
 )
