--- openssh-3.3p1/scp.c.orig	Fri Jun 21 02:41:52 2002
+++ openssh-3.3p1/scp.c	Sun Jun 23 21:29:30 2002
@@ -233,9 +233,11 @@
 	addargs(&args, "-oClearAllForwardings yes");
 
 	fflag = tflag = 0;
-	while ((ch = getopt(argc, argv, "dfprtvBCc:i:P:q46S:o:F:")) != -1)
+	while ((ch = getopt(argc, argv, "dfprtvBCc:i:P:q1246S:o:F:")) != -1)
 		switch (ch) {
 		/* User-visible flags. */
+		case '1':
+		case '2':
 		case '4':
 		case '6':
 		case 'C':
@@ -955,7 +957,7 @@
 usage(void)
 {
 	(void) fprintf(stderr,
-	    "usage: scp [-pqrvBC46] [-F config] [-S program] [-P port]\n"
+	    "usage: scp [-pqrvBC1246] [-F config] [-S program] [-P port]\n"
 	    "           [-c cipher] [-i identity] [-o option]\n"
 	    "           [[user@]host1:]file1 [...] [[user@]host2:]file2\n");
 	exit(1);
--- openssh-2.9.9p2/scp.1.orig	Tue Sep 18 07:56:57 2001
+++ openssh-2.9.9p2/scp.1	Fri Sep 28 05:30:54 2001
@@ -19,7 +19,7 @@
 .Nd secure copy (remote file copy program)
 .Sh SYNOPSIS
 .Nm scp
-.Op Fl pqrvBC46
+.Op Fl pqrvBC1246
 .Op Fl F Ar ssh_config
 .Op Fl S Ar program
 .Op Fl P Ar port
@@ -125,6 +125,14 @@
 command-line flag.  For example, forcing the use of protocol
 version 1 is specified using
 .Ic scp -oProtocol=1 .
+.It Fl 1
+Forces
+.Nm
+to use SSH1 only.
+.It Fl 2
+Forces
+.Nm
+to use SSH2 only.
 .It Fl 4
 Forces
 .Nm
