--- openssh-6.9p1/regress/netcat.c.orig	2015-07-01 04:35:31.000000000 +0200
+++ openssh-6.9p1/regress/netcat.c	2015-07-03 17:23:33.544777525 +0200
@@ -738,7 +738,7 @@ local_listen(char *host, char *port, str
 #ifdef SO_REUSEPORT
 		ret = setsockopt(s, SOL_SOCKET, SO_REUSEPORT, &x, sizeof(x));
 		if (ret == -1)
-			err(1, "setsockopt SO_REUSEPORT");
+			warn("setsockopt SO_REUSEPORT");
 #endif
 #ifdef SO_REUSEADDR
 		ret = setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &x, sizeof(x));

