diff -bruN openssh-2.1.0/Makefile.in openssh-2.1.0.new/Makefile.in
--- openssh-2.1.0/Makefile.in	Tue May  9 06:28:55 2000
+++ openssh-2.1.0.new/Makefile.in	Wed May 10 19:16:45 2000
@@ -23,6 +23,7 @@
 PATHS=-DETCDIR=\"$(sysconfdir)\" -DSSH_PROGRAM=\"$(SSH_PROGRAM)\" -DSSH_ASKPASS_DEFAULT=\"$(ASKPASS_PROGRAM)\"
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -71,7 +72,7 @@
 	$(LD) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
 
 sshd: libssh.a	libopenbsd-compat.a $(SSHDOBJS)
-	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
+	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) $(LIBWRAP)
 
 scp: libopenbsd-compat.a libssh.a scp.o
 	$(LD) -o $@ scp.o $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) 
diff -bruN openssh-2.1.0/configure.in openssh-2.1.0.new/configure.in
--- openssh-2.1.0/configure.in	Mon May  8 12:49:37 2000
+++ openssh-2.1.0.new/configure.in	Wed May 10 20:23:33 2000
@@ -757,7 +757,10 @@
 	[
 		if test "x$withval" != "xno" ; then
 			saved_LIBS="$LIBS"
-			LIBS="$LIBS -lwrap"
+			LIBWRAP="-lwrap"
+			AC_SUBSCT(LIBWRAP)
+			AC_DEFINE(LIBWRAP)
+			LIBS="$saved_LIBS"
 			AC_MSG_CHECKING(for libwrap)
 			AC_TRY_LINK(
 				[
