diff -Nru openssh-2.1.1p2/Makefile.in openssh-2.1.1p2.new/Makefile.in
--- openssh-2.1.1p2/Makefile.in	Thu Jul  6 05:30:34 2000
+++ openssh-2.1.1p2.new/Makefile.in	Thu Jul  6 05:29:57 2000
@@ -23,6 +23,7 @@
 PATHS=-DETCDIR=\"$(sysconfdir)\" -DSSH_PROGRAM=\"$(SSH_PROGRAM)\" -DSSH_ASKPASS_DEFAULT=\"$(ASKPASS_PROGRAM)\"
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -72,7 +73,7 @@
 	$(LD) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
 
 sshd: libssh.a	libopenbsd-compat.a $(SSHDOBJS)
-	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
+	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) $(LIBWRAP)
 
 scp: libopenbsd-compat.a libssh.a scp.o
 	$(LD) -o $@ scp.o $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) 
diff -Nru openssh-2.1.1p2/configure.in openssh-2.1.1p2.new/configure.in
--- openssh-2.1.1p2/configure.in	Thu Jul  6 05:30:34 2000
+++ openssh-2.1.1p2.new/configure.in	Thu Jul  6 05:31:54 2000
@@ -929,8 +929,11 @@
 	[  --with-tcp-wrappers     Enable tcpwrappers support],
 	[
 		if test "x$withval" != "xno" ; then
+			# we need to add -lwrap to LIBS only for configure
+			# testing, and then remove
 			saved_LIBS="$LIBS"
-			LIBS="$LIBS -lwrap"
+			LIBWRAP="-lwrap"
+			LIBS="$LIBS $LIBWRAP"
 			AC_MSG_CHECKING(for libwrap)
 			AC_TRY_LINK(
 				[
@@ -941,12 +944,14 @@
 				[
 					AC_MSG_RESULT(yes)
 					AC_DEFINE(LIBWRAP)
+					AC_SUBST(LIBWRAP)
 					TCPW_MSG="yes" 
 				],
 				[
 					AC_MSG_ERROR([*** libwrap missing])
 				]
 			)
+			LIBS="$saved_LIBS"
 		fi
 	]
 )
