--- openssh-1.2.2p1/Makefile.in.wiget	Fri Mar  3 12:13:52 2000
+++ openssh-1.2.2p1/Makefile.in	Tue Mar  7 15:44:57 2000
@@ -23,6 +23,7 @@
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 EXTRA_TARGETS=@GNOME_ASKPASS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -62,7 +63,7 @@
 	$(CC) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh $(LIBS)
 
 sshd: libssh.a	$(SSHDOBJS)
-	$(CC) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh $(LIBS)
+	$(CC) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh $(LIBWRAP) $(LIBS)
 
 scp: libssh.a scp.o
 	$(CC) -o $@ scp.o $(LDFLAGS) -lssh $(LIBS) 
--- openssh-1.2.2p1/configure.in.wiget	Sun Mar  5 07:02:46 2000
+++ openssh-1.2.2p1/configure.in	Tue Mar  7 15:50:14 2000
@@ -711,6 +711,9 @@
 				[
 					AC_MSG_RESULT(yes)
 					AC_DEFINE(LIBWRAP)
+					LIBWRAP="-lwrap"
+					AC_SUBST(LIBWRAP)
+					LIBS="$saved_LIBS"
 				],
 				[
 					AC_MSG_RESULT(no)
