diff -ur openssh-9.5p1.org/ssh_config openssh-9.5p1/ssh_config
--- openssh-9.5p1.org/ssh_config	2023-10-04 06:34:10.000000000 +0200
+++ openssh-9.5p1/ssh_config	2023-11-28 09:12:00.249971177 +0100
@@ -44,3 +44,6 @@
 #   ProxyCommand ssh -q -W %h:%p gateway.example.com
 #   RekeyLimit 1G 1h
 #   UserKnownHostsFile ~/.ssh/known_hosts.d/%k
+
+# Put your local config in *.conf files
+Include /etc/ssh/ssh_config.d/*.conf
diff -ur openssh-9.5p1.org/sshd_config openssh-9.5p1/sshd_config
--- openssh-9.5p1.org/sshd_config	2023-10-04 06:34:10.000000000 +0200
+++ openssh-9.5p1/sshd_config	2023-11-28 09:12:18.119971176 +0100
@@ -114,3 +114,6 @@
 #	AllowTcpForwarding no
 #	PermitTTY no
 #	ForceCommand cvs server
+
+# Put your local config in *.conf files
+Include /etc/ssh/sshd_config.d/*.conf
