diff -Naur openssh-3.1p1/configure.ac openssh-3.1p1-p/configure.ac
--- openssh-3.1p1/configure.ac	Mon Mar 25 12:19:11 2002
+++ openssh-3.1p1-p/configure.ac	Mon Mar 25 12:19:42 2002
@@ -121,7 +121,6 @@
 	no_dev_ptmx=1
 	check_for_libcrypt_later=1
 	check_for_openpty_ctty_bug=1
-	AC_DEFINE(DONT_TRY_OTHER_AF)
 	AC_DEFINE(PAM_TTY_KLUDGE)
 	AC_DEFINE(LOCKED_PASSWD_PREFIX, "!!")
 	AC_DEFINE(SPT_TYPE,SPT_REUSEARGV)
