diff -Naur openssh-2.1.1p1/Makefile.in openssh-2.1.1p1-p/Makefile.in
--- openssh-2.1.1p1/Makefile.in	Sat Jun  3 16:57:40 2000
+++ openssh-2.1.1p1-p/Makefile.in	Mon Jun 12 16:33:55 2000
@@ -23,6 +23,7 @@
 PATHS=-DETCDIR=\"$(sysconfdir)\" -DSSH_PROGRAM=\"$(SSH_PROGRAM)\" -DSSH_ASKPASS_DEFAULT=\"$(ASKPASS_PROGRAM)\"
 CFLAGS=@CFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
+LIBWRAP=@LIBWRAP@
 AR=@AR@
 RANLIB=@RANLIB@
 INSTALL=@INSTALL@
@@ -72,7 +73,7 @@
 	$(LD) -o $@ $(SSHOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
 
 sshd: libssh.a	libopenbsd-compat.a $(SSHDOBJS)
-	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
+	$(LD) -o $@ $(SSHDOBJS) $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) $(LIBWRAP)
 
 scp: libopenbsd-compat.a libssh.a scp.o
 	$(LD) -o $@ scp.o $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS) 
diff -Naur openssh-2.1.1p1/configure.in openssh-2.1.1p1-p/configure.in
--- openssh-2.1.1p1/configure.in	Fri Jun  9 03:58:35 2000
+++ openssh-2.1.1p1-p/configure.in	Mon Jun 12 16:33:55 2000
@@ -814,8 +814,11 @@
 	[  --with-tcp-wrappers     Enable tcpwrappers support],
 	[
 		if test "x$withval" != "xno" ; then
+			# we need to add -lwrap to LIBS only for configure
+			# testing, and then remove
 			saved_LIBS="$LIBS"
-			LIBS="$LIBS -lwrap"
+			LIBWRAP="-lwrap"
+			LIBS="$LIBS $LIBWRAP"
 			AC_MSG_CHECKING(for libwrap)
 			AC_TRY_LINK(
 				[
@@ -826,6 +829,7 @@
 				[
 					AC_MSG_RESULT(yes)
 					AC_DEFINE(LIBWRAP)
+					AC_SUBST(LIBWRAP)
 				],
 				[
 					AC_MSG_RESULT(no)
@@ -833,6 +837,7 @@
 					LIBS="$saved_LIBS"
 				]
 			)
+			LIBS="$saved_LIBS"
 		fi
 	]
 )
